### ISSUE TITLE
XMap raises error when used inside jax.value_and_grad

### ISSUE THREAD

**Issue by Conchylicultor** [google/jax#5848]
Reproduction instructions:

```python
def grid_sample(img, coords) -> f32['b n c']:
  return jax.experimental.maps.xmap(
      functools.partial(
          jax.scipy.ndimage.map_coordinates,
          order=1, 
      ),
      in_axes=[['b', ...], ['b', ...]],
      out_axes=['b', ...],
  )(img, coords)

loss_and_grad_fn = jax.value_and_grad(grid_sample)

loss_and_grad_fn(jnp.zeros((2, 5, 5)), jnp.zeros((2, 2, 2)))
```

Raises:

```
NotImplementedError: <class 'jax.interpreters.ad.JVPTrace'> must override process_xmap to handle xmap
```

If I replace `xmap` by `vmap`, it works fine.

<details>
  <summary>Full stacktrace</summary>

```
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-66-c2360840b257> in <module>()
     11 loss_and_grad_fn = jax.value_and_grad(grid_sample)
     12 
---> 13 loss_and_grad_fn(jnp.zeros((2, 5, 5, 1)), jnp.zeros((2, 2, 2)),)

11 frames
google3/third_party/py/jax/api.py in value_and_grad_f(*args, **kwargs)
    803     tree_map(partial(_check_input_dtype_grad, holomorphic, allow_int), dyn_args)
    804     if not has_aux:
--> 805       ans, vjp_py = _vjp(f_partial, *dyn_args)
    806     else:
    807       ans, vjp_py, aux = _vjp(f_partial, *dyn_args, has_aux=True)

google3/third_party/py/jax/api.py in _vjp(fun, has_aux, *primals)
   1876   if not has_aux:
   1877     flat_fun, out_tree = flatten_fun_nokwargs(fun, in_tree)
-> 1878     out_primal, out_vjp = ad.vjp(flat_fun, primals_flat)
   1879     out_tree = out_tree()
   1880   else:

google3/third_party/py/jax/interpreters/ad.py in vjp(traceable, primals, has_aux)
    112 def vjp(traceable, primals, has_aux=False):
    113   if not has_aux:
--> 114     out_primals, pvals, jaxpr, consts = linearize(traceable, *primals)
    115   else:
    116     out_primals, pvals, jaxpr, consts, aux = linearize(traceable, *primals, has_aux=True)

google3/third_party/py/jax/interpreters/ad.py in linearize(traceable, *primals, **kwargs)
     99   _, in_tree = tree_flatten(((primals, primals), {}))
    100   jvpfun_flat, out_tree = flatten_fun(jvpfun, in_tree)
--> 101   jaxpr, out_pvals, consts = pe.trace_to_jaxpr(jvpfun_flat, in_pvals)
    102   out_primals_pvals, out_tangents_pvals = tree_unflatten(out_tree(), out_pvals)
    103   assert all(out_primal_pval.is_known() for out_primal_pval in out_primals_pvals)

google3/third_party/py/jax/interpreters/partial_eval.py in trace_to_jaxpr(fun, pvals, instantiate)
    504   with core.new_main(JaxprTrace) as main:
    505     fun = trace_to_subjaxpr(fun, main, instantiate)
--> 506     jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
    507     assert not env
    508     del main, fun, env

google3/third_party/py/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    164 
    165     try:
--> 166       ans = self.f(*args, **dict(self.params, **kwargs))
    167     except:
    168       # Some transformations yield from inside context managers, so we have to

<ipython-input-66-c2360840b257> in grid_sample(img, coords)
      7       in_axes=[['b', ...], ['b', ...]],
      8       out_axes=['b', ...],
----> 9   )(img, coords)
     10 
     11 loss_and_grad_fn = jax.value_and_grad(grid_sample)

google3/third_party/py/jax/experimental/maps.py in fun_mapped(*args)
    504       axis_resources=frozen_axis_resources,
    505       resource_env=resource_env,
--> 506       backend=backend)
    507     if has_output_rank_assertions:
    508       for out, spec in zip(out_flat, out_axes_thunk()):

google3/third_party/py/jax/experimental/maps.py in bind(self, fun, *args, **params)
    619   def bind(self, fun, *args, **params):
    620     assert len(params['in_axes']) == len(args)
--> 621     return core.call_bind(self, fun, *args, **params)  # type: ignore
    622 
    623   def process(self, trace, fun, tracers, params):

google3/third_party/py/jax/core.py in call_bind(primitive, fun, *args, **params)
   1401   tracers = map(top_trace.full_raise, args)
   1402   with maybe_new_sublevel(top_trace):
-> 1403     outs = primitive.process(top_trace, fun, tracers, params)
   1404   return map(full_lower, apply_todos(env_trace_todo(), outs))
   1405 

google3/third_party/py/jax/experimental/maps.py in process(self, trace, fun, tracers, params)
    622 
    623   def process(self, trace, fun, tracers, params):
--> 624     return trace.process_xmap(self, fun, tracers, params)
    625 
    626   def post_process(self, trace, out_tracers, params):

google3/third_party/py/jax/experimental/maps.py in _process_xmap_default(self, call_primitive, f, tracers, params)
    630 core.EvalTrace.process_xmap = core.EvalTrace.process_call  # type: ignore
    631 def _process_xmap_default(self, call_primitive, f, tracers, params):
--> 632   raise NotImplementedError(f"{type(self)} must override process_xmap to handle xmap")
    633 core.Trace.process_xmap = _process_xmap_default  # type: ignore
    634 

NotImplementedError: <class 'jax.interpreters.ad.JVPTrace'> must override process_xmap to handle xmap
```
</details>

**Comment by merrymercy** [comment]
I got a similar error. Are there any examples to use grad with xmap? Is it supported now?

**Comment by rajasekharporeddy** [comment]
Hi @Conchylicultor 

Looks like this issue has been resolved in later versions of JAX (from JAX 0.3.1). I tested the provided repro code on colab CPU with JAX version 0.4.26. It now produces a `TypeError: Gradient only defined for scalar-output functions. Output had shape: (2, 2)`, since the `grid_sample` function returns an array of 2 dimensions and `grad` is defined only for `scalar outputs`. 

```python
import jax
import jax.numpy as jnp
import functools
from jax.experimental.maps import xmap


def grid_sample(img, coords):
  return xmap(
      functools.partial(
          jax.scipy.ndimage.map_coordinates,
          order=1,
      ),
      in_axes=[['b', ...], ['b', ...]],
      out_axes=['b', ...],
  )(img, coords)

loss_and_grad_fn = jax.value_and_grad(grid_sample)

print(grid_sample(jnp.zeros((2, 5, 5)), jnp.zeros((2, 2, 2))))

loss_and_grad_fn(jnp.zeros((2, 5, 5)), jnp.zeros((2, 2, 2)))
```
Output:
```
[[0. 0.]
 [0. 0.]]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-3-c1e8ceaf7dd5>](https://localhost:8080/#) in <cell line: 21>()
     19 print(grid_sample(jnp.zeros((2, 5, 5)), jnp.zeros((2, 2, 2))))
     20 
---> 21 loss_and_grad_fn(jnp.zeros((2, 5, 5)), jnp.zeros((2, 2, 2)))

    [... skipping hidden 2 frame]

[/usr/local/lib/python3.10/dist-packages/jax/_src/api.py](https://localhost:8080/#) in _check_scalar(x)
    739     if isinstance(aval, ShapedArray):
    740       if aval.shape != ():
--> 741         raise TypeError(msg(f"had shape: {aval.shape}"))
    742     else:
    743       raise TypeError(msg(f"had abstract value {aval}"))

TypeError: Gradient only defined for scalar-output functions. Output had shape: (2, 2).
```

If we change the code to make `grid_sample` function to return scalar output, then it works without any error.

```python
import jax
import jax.numpy as jnp
import functools
from jax.experimental.maps import xmap


def grid_sample(img, coords):
  return jnp.sum(xmap(
      functools.partial(
          jax.scipy.ndimage.map_coordinates,
          order=1,
      ),
      in_axes=[['b', ...], ['b', ...]],
      out_axes=['b', ...],
  )(img, coords))

loss_and_grad_fn = jax.value_and_grad(grid_sample)

print(grid_sample(jnp.zeros((2, 5, 5)), jnp.zeros((2, 2, 2))))

loss_and_grad_fn(jnp.zeros((2, 5, 5)), jnp.zeros((2, 2, 2)))
```
Output:
```
0.0
(Array(0., dtype=float32),
 Array([[[2., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],
 
        [[2., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]], dtype=float32))
```
Attaching the [gist](https://colab.sandbox.google.com/gist/rajasekharporeddy/78058b7d8b37acb3a6e055fd58827dbf/-5848.ipynb) for reference.

Thank you.

**Comment by apaszke** [comment]
xmap was removed from JAX, so this is obsolete
